File: software/examples/go/example_callback.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
	"github.com/Tinkerforge/go-api-bindings/stepper_brick"
	"math/rand"
	"time"
)

const ADDR string = "localhost:4223"
const UID string = "XXYYZZ" // Change XXYYZZ to the UID of your Stepper Brick.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	stepper, _ := stepper_brick.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	rng := rand.New(rand.NewSource(time.Now().UnixNano()))

	stepper.RegisterPositionReachedCallback(func(position int32) {
		steps := int32(0)
		if rng.Int31()%2 == 0 {
			steps = rng.Int31n(4001) + 1000
			fmt.Printf("Driving forward: %d steps\n", steps)
		} else {
			steps = -(rng.Int31n(4001) + 1000)
			fmt.Printf("Driving backward: %d steps\n", steps)
		}

		vel := uint16(rng.Int31n(1800) + 200)
		acc := uint16(rng.Int31n(900) + 100)
		dec := uint16(rng.Int31n(900) + 100)

		fmt.Printf("Configuration: (vel, acc, dec): (%d, %d, %d)\n", vel, acc, dec)
		stepper.SetSpeedRamping(acc, dec)
		stepper.SetMaxVelocity(vel)
		stepper.SetSteps(steps)
	})

	stepper.Enable()    // Enable motor power
	stepper.SetSteps(1) // Drive one step forward to get things going

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

	// Stop motor before disabling motor power
	stepper.Stop() // Request motor stop
	stepper.SetSpeedRamping(500,
		5000) // Fast deacceleration (5000 steps/s^2) for stopping
	time.Sleep(400 * time.Millisecond) // Wait for motor to actually stop: max velocity (2000 steps/s) / decceleration (5000 steps/s^2) = 0.4 s
	stepper.Disable()                  // Disable motor power
}

File: software/examples/rust/example_callback.rs
use rand::{thread_rng, Rng};
use std::{error::Error, io, thread};
use tinkerforge::{ip_connection::IpConnection, stepper_brick::*};

const HOST: &str = "127.0.0.1";
const PORT: u16 = 4223;
const UID: &str = "XXYYZZ"; // Change XXYYZZ to the UID of your Stepper Brick

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection
    let stepper = StepperBrick::new(UID, &ipcon); // Create device object

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd
                                          // Don't use device before ipcon is connected

    let position_reached_receiver = stepper.get_position_reached_callback_receiver();

    // Spawn thread to handle received callback messages. This thread ends when the stepper
    // is dropped, so there is no need for manual cleanup.
    let stepper_copy = stepper.clone(); //Device objects don't implement Sync, so they can't be shared between threads (by reference). So clone the device and move the copy.
    thread::spawn(move || {
        let mut rng = thread_rng();
        for _position_reached in position_reached_receiver {
            let steps = if rng.gen() {
                let steps = rng.gen_range(1000, 5001); // steps (forward)
                println!("Driving forward: {} steps", steps);
                steps
            } else {
                let steps = rng.gen_range(-5000, -999); // steps (backward)
                println!("Driving backward: {} steps", steps);
                steps
            };

            let vel = rng.gen_range(200, 2001); // steps/s
            let acc = rng.gen_range(100, 1001); // steps/s^2
            let dec = rng.gen_range(100, 1001); // steps/s^2

            println!("Configuration (vel, acc, dec): ({}, {}, {})", vel, acc, dec);

            stepper_copy.set_speed_ramping(acc, dec);
            stepper_copy.set_max_velocity(vel);
            stepper_copy.set_steps(steps);
        }
    });

    stepper.enable(); // Enable motor power
    stepper.set_steps(1); // Drive one step forward to get things going

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    stepper.disable();
    ipcon.disconnect();
    Ok(())
}
